### DISCUSSION TITLE
How to refetch blog post data on route change

### DISCUSSION THREAD

**Discussion by orafal-dev** [nuxt/nuxt#16385] [Questions] [answered]
I'm using `useFetch` to get data from Strapi, data is fetching just fine. But whenever I switch route to another blog post, it requires full browser refresh to get new data of the post I switched to.

I've thought of key'ing route as you would normally do with `<RouterView :key="route.fullPath" />` (tried that with `definePageMeta`). That didn't help me at all. Blog post file is named `pages/blog/articles/[slug].vue`

Here's quick snippet with fetching and tried it even with `watch()`.

```js
definePageMeta({
  key: (route) => route.params.slug,
});

const config = useRuntimeConfig();
const { data, refresh } = await useFetch(
  `${config.public.strapiBaseUrl}/blog-posts/${route.params.slug}`
);

watch(
  () => route.params.slug,
  () => {
    refresh();
  }
);
```

What are your solutions to this case?

**Comment by Zerro97** [comment, marked as answer]
Hi! Maybe including key in a useFetch parameter options could help. 

In my project I use [NuxtStrapi](https://strapi.nuxtjs.org/) with useAsyncData and provide a unique key for each request. I think it could be same for [useFetch](https://v3.nuxtjs.org/api/composables/use-fetch#params) as well

```javascript
const { find } = useStrapi();
const { data, refresh } = await useAsyncData(
    'unique key', // <--- you can give a key here
    async () => await find(`${config.public.strapiBaseUrl}/blog-posts/${route.params.slug}`)
);
```

**Reply by Zerro97** [reply]
Just to add on, I think the reason why it doesn't refetch is that the key used in useFetch are same across different post pages (since key generation depends on file name and line number):

> **key:** a unique key to ensure that data fetching can be properly de-duplicated across requests. If you do not provide a key, then a key that is unique to the file name and line number of the instance of useAsyncData will be generated for you.

[source](https://v3.nuxtjs.org/api/composables/use-async-data#params)

**Reply by TheAlexLichter** [reply]
Yes, ideally you pick a key that takes route params necessary for the fetch into account. Ideally this will be automatically the case via https://github.com/nuxt/framework/pull/6632

**Reply by orafal-dev** [reply]
Thanks guys! That's what solved my issue. Didn't quite understand it at first, but now it's clear 😄
